src/wasm/wasm-binary.cpp

```diff
@@ -945,6 +945,15 @@ void WasmBinaryBuilder::read() {
 
     auto oldPos = pos;
 
+    // almost no sections can appear more than once
+    if (seenSections.count(BinaryConsts::Section(sectionCode)) > 0) {
+      if (sectionCode != BinaryConsts::Section::User && sectionCode != BinaryConsts::Section::Code) {
+        throw ParseException("section seen more than once: " + std::to_string(sectionCode));
+      }
+    } else {
+      seenSections.insert(BinaryConsts::Section(sectionCode));
```

Comment:
Well, it could be even simpler than this, since sections have to appear in order as well. So you can just keep an int tracking the last seen section and compare against that.